### PR TITLE
[20.09] pythonPackages.pymc3: fix build, init pythonPackages.fastprogress at 1.0.0

### PR DESCRIPTION
**Pull request by risicle**
###### Motivation for this change
ZHF: #97479

This is a modified & combined backport of #99587 and #100867.

Because 20.09 is missing #97597 and #93560 (and I think including them at this stage might be a bit risky), importing/using `theano` from within a nix build environment (i.e. with `NIX_ENFORCE_PURITY=1`) will only succeed if the tmpdir used is actually literally `/tmp`, which you can't really do from a sandboxed build. So this uses `libredirect` to convince the compiler wrapper we are actually working with `/tmp`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by risicle** [comment]
@ofborg build python37Packages.pymc3

File: pkgs/development/python-modules/Theano/default.nix [hunk 1]
```diff
@@ -68,9 +69,22 @@ in buildPythonPackage rec {
       --replace 'StrParam(default_dnn_base_path)' 'StrParam('\'''${cudnn}'\''')'
   '';
 
-  preCheck = ''
+  # needs to be postFixup so it runs before pythonImportsCheck even when
+  # doCheck = false (meaning preCheck would be disabled).
+  # this branch is missing #97597 (and its predecessor #93560) meaning only
+  # "/tmp" is exempt from NIX_ENFORCE_PURITY's objections when theano is
+  # imported from within a nix build environment. therefore use libredirect
+  # to convince the wrapper we are actually accessing "/tmp"
```
Comment (das-g) [inline pkgs/development/python-modules/Theano/default.nix:77]: Begin sentences in comments with a capital letter. Also I think a comma more is needed:
```suggestion
  # Needs to be postFixup so it runs before pythonImportsCheck even when
  # doCheck = false (meaning preCheck would be disabled).
  # This branch is missing #97597 (and its predecessor #93560), meaning only
  # "/tmp" is exempt from NIX_ENFORCE_PURITY's objections when theano is
  # imported from within a nix build environment. Therefore use libredirect
  # to convince the wrapper we are actually accessing "/tmp".
```

File: pkgs/development/python-modules/pymc3/default.nix [hunk 2]
```diff
@@ -52,11 +57,24 @@ buildPythonPackage rec {
...
+  pythonImportsCheck = [ "pymc3" ];
 
   # For some reason tests are run as a part of the *install* phase if enabled.
   # Theano writes compiled code to ~/.theano hence we set $HOME.
-  preInstall = "export HOME=$(mktemp -d)";
-  postInstall = "rm -rf $HOME";
+  # This branch is missing #97597 (and its predecessor #93560) meaning only
+  # "/tmp" is exempt from NIX_ENFORCE_PURITY's objections when theano is
+  # imported from within a nix build environment. therefore use libredirect
+  # to convince the wrapper we are actually accessing "/tmp"
```
Comment (das-g) [inline pkgs/development/python-modules/pymc3/default.nix:67]: Same here:
```suggestion
  # This branch is missing #97597 (and its predecessor #93560), meaning only
  # "/tmp" is exempt from NIX_ENFORCE_PURITY's objections when theano is
  # imported from within a nix build environment. Therefore use libredirect
  # to convince the wrapper we are actually accessing "/tmp".
```

**Review by das-g — CHANGES_REQUESTED** [review]
Please use correct grammar in comments.

**Comment by risicle** [comment]
Er.. not sure if you're serious, but ok...

**Comment by das-g** [comment]
Result of `nixpkgs-review pr 100955` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>python27Packages.Lasagne</li>
    <li>python27Packages.Theano (python27Packages.TheanoWithoutCuda)</li>
    <li>python27Packages.TheanoWithCuda</li>
    <li>python37Packages.Theano (python37Packages.TheanoWithoutCuda)</li>
    <li>python37Packages.TheanoWithCuda</li>
    <li>python37Packages.fastprogress</li>
    <li>python37Packages.pymc3</li>
    <li>python38Packages.Theano (python38Packages.TheanoWithoutCuda)</li>
    <li>python38Packages.TheanoWithCuda</li>
    <li>python38Packages.fastprogress</li>
  </ul>
</details>

**Review by das-g — APPROVED** [review]

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 100955 [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>10 packages built:</summary>  <ul>    <li>python27Packages.Lasagne</li>    <li>python27Packages.Theano (python27Packages.TheanoWithoutCuda)</li>    <li>python27Packages.TheanoWithCuda</li>    <li>python37Packages.Theano (python37Packages.TheanoWithoutCuda)</li>    <li>python37Packages.TheanoWithCuda</li>    <li>python37Packages.fastprogress</li>    <li>python37Packages.pymc3</li>    <li>python38Packages.Theano (python38Packages.TheanoWithoutCuda)</li>    <li>python38Packages.TheanoWithCuda</li>    <li>python38Packages.fastprogress</li>  </ul></details>

**Comment by twhitehead** [comment]
@risicle BTW, just wanted to say thanks for working this all out. Have been reading through the various threads and can see this was quite a mess for you to straighten out. :+1:

**Comment by risicle** [comment]
No problem - you got the _real_ fix in place.

**Pull request merged**